### PR TITLE
Openstack defaulting

### LINKED ISSUE #45: Defaulting for Openstack
Usage of the Openstack provider would be easier if there was defaulting for

* availabilityZone
* Region
* Network
* Subnet
* FloatingIPPool

To achieve this, the machine-controller should request a list of the given resource, check if there is exactly one and if yes default to that.

### PR DESCRIPTION
**Pull request by alvaroaleman**
**What this PR does / why we need it**:

Adds some defaulting logic for the Openstack provider

**Which issue(s) this PR fixes** 
Fixes #45

### CODE REVIEW

File: pkg/controller/machine.go [hunk 1]
```diff
@@ -168,12 +168,16 @@ func (c *Controller) processNextWorkItem() bool {
 	return true
 }
 
+func (c *Controller) updateMachine(machine *machinev1alpha1.Machine) (*machinev1alpha1.Machine, error) {
+	machine.Status.LastUpdated = metav1.Now()
+	return c.machineClient.MachineV1alpha1().Machines().Update(machine)
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: very good catch!

File: pkg/cloudprovider/provider/openstack/provider.go [hunk 2]
```diff
@@ -77,6 +80,24 @@ func getConfig(s runtime.RawExtension) (*Config, *providerconfig.Config, error)
 	return &c, &pconfig, err
 }
 
+func setProviderConfig(config *Config, spec v1alpha1.MachineSpec) (v1alpha1.MachineSpec, error) {
+	var providerConfig map[string]interface{}
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/openstack/provider.go]: Instead unmarshal into the correct struct: `providerconfig.Config`

File: pkg/cloudprovider/provider/openstack/provider.go [hunk 3]
```diff
@@ -87,7 +108,124 @@ func getClient(c *Config) (*gophercloud.ProviderClient, error) {
...
+			for _, network := range networks {
+				for _, subnet := range network.Subnets {
+					_, err := getSubnet(client, c.Region, subnet)
+					if err == errNotFound {
+						continue
+					} else if err != nil {
+						return spec, changed, fmt.Errorf("Failed to retrieve subnet: '%v'", err)
+					}
+					candidates = append(candidates, network)
+					continue NetworkLoop
```
Comment (mrIncompetent) [inline pkg/cloudprovider/provider/openstack/provider.go]: Can we put the logic for retrieving the default network into a separate function?

File: pkg/controller/machine.go [hunk 4]
```diff
@@ -264,9 +267,22 @@ func (c *Controller) syncHandler(key string) error {
 	providerInstance, err := prov.Get(machine)
 	if err != nil {
 		if err == cloudprovidererrors.ErrInstanceNotFound {
-			if err := prov.Validate(machine.Spec); err != nil {
-				if machine, err = c.updateMachineError(machine, machinev1alpha1.InvalidConfigurationMachineError, err.Error()); err != nil {
-					return fmt.Errorf("failed to update machine error after failed validation: %v", err)
+			defaultedMachine, changed, err := prov.AddDefaults(machine.Spec)
+			if err != nil {
+				return fmt.Errorf("Failed to add defaults to machine: '%v'", err)
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: errors should be lowercase

File: pkg/controller/machine.go [hunk 5]
```diff
@@ -264,9 +267,22 @@ func (c *Controller) syncHandler(key string) error {
 	providerInstance, err := prov.Get(machine)
 	if err != nil {
 		if err == cloudprovidererrors.ErrInstanceNotFound {
-			if err := prov.Validate(machine.Spec); err != nil {
-				if machine, err = c.updateMachineError(machine, machinev1alpha1.InvalidConfigurationMachineError, err.Error()); err != nil {
-					return fmt.Errorf("failed to update machine error after failed validation: %v", err)
+			defaultedMachine, changed, err := prov.AddDefaults(machine.Spec)
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: defaultedMachineSpec

File: pkg/controller/machine.go [hunk 6]
```diff
@@ -264,9 +267,22 @@ func (c *Controller) syncHandler(key string) error {
...
+			if err != nil {
+				return fmt.Errorf("Failed to add defaults to machine: '%v'", err)
+			}
+			if changed {
+				glog.V(4).Infof("Updating machine with defaults...")
+				machine.Spec = defaultedMachine
+				machine, err = c.updateMachine(machine)
+				if err != nil {
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: inline update

File: pkg/controller/machine.go [hunk 7]
```diff
@@ -264,9 +267,22 @@ func (c *Controller) syncHandler(key string) error {
...
+					return fmt.Errorf("Failed to update machine after adding defaults: '%v'", err)
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: errors should be lowercase

File: pkg/controller/machine.go [hunk 8]
```diff
@@ -264,9 +267,22 @@ func (c *Controller) syncHandler(key string) error {
...
+				}
+				glog.V(4).Infof("Successfully updated machine with defaults!")
+			}
+			if err = prov.Validate(machine.Spec); err != nil {
+				if _, errNested := c.updateMachineError(machine, machinev1alpha1.InvalidConfigurationMachineError, err.Error()); errNested != nil {
+					return fmt.Errorf("failed to update machine error after failed validation: %v", errNested)
```
Comment (mrIncompetent) [inline pkg/controller/machine.go]: errors should be lowercase

**Comment by alvaroaleman** [comment]
@mrIncompetent  I think I've addressed all your points, PTAL.

**Pull request merged**